academic-services/knowledge-exploration-service/linkPrivateLibraryData.ps1

```diff
@@ -0,0 +1,191 @@
+    $privateLibraryPaper | Add-Member -MemberType NoteProperty -Name 'EstimatedCitationCount' -Value  $magEntity.ECC
+
+    # add fields of study 
+    $fieldsOfStudy = @()
+    foreach($fos in $magEntity.F)
+    {
+        $fieldsOfStudy += 
+        @{ 
+            OriginalName = $fos.DFN;
+            Name = $fos.FN;
+        }
+    }
+    $privateLibraryPaper | Add-Member -MemberType NoteProperty -Name 'FieldsOfStudy' -Value  $fieldsOfStudy
+
+    # add structured author affiliations
+    $authorAffiliations = @()
+    foreach ($authorAffiliation in $magEntity.AA)
+    {
+        $temp = @{
+            OriginalAuthorName = $authorAffiliation.DAuN;
+            AuthorName = $authorAffiliation.AuN;
+        }
+
+        if ($null -ne $authorAffiliation.DAfN)
+        {
+            $temp.OriginalAffiliationName = $authorAffiliation.DAfN;
+        }
+        if ($null -ne $authorAffiliation.AfN)
+        {
+            $temp.AffiliationName = $authorAffiliation.AfN;
+        }
+        if ($null -ne $authorAffiliation.S)
+        {
+            $temp.Sequence = $authorAffiliation.S;
+        }
+        $authorAffiliations += $temp;
+    }
+    $privateLibraryPaper | Add-Member -MemberType NoteProperty -Name 'AuthorAffiliations' -Value  $authorAffiliations
+
+    # add abstract by reconstructing it from inverted abstract
+    if ($null -ne $magEntity.IA)
+    {
+        $abstractWords = [string[]]::new($magEntity.IA.IndexLength)
+        foreach ($pair in $magEntity.IA.InvertedIndex.GetEnumerator())
+        {
+            $abstractWord = $pair.Name
+            foreach ($index in $pair.Value)
+            {
+                $abstractWords[$index] = $abstractWord
+            }
+        }
+        $abstract = [string]::Join(" ", $abstractWords)
+        $privateLibraryPaper | Add-Member -MemberType NoteProperty -Name 'OriginalAbstract' -Value  $abstract
+    }
+}
+
+function ConvertTo-NormalizedStr {
+    param($str)
+    # replace all non-alphanumeric characters with a space
+    $str = [Regex]::Replace($str, "[^\d\w]", " ")
+    # reduce all instances of 1+ spaces to a single space
+    $str = [Regex]::Replace($str, "\s+", " ")
+    # convert to lower case
+    $str = $str.ToLower()
+    return $str
+}
+
+function ConvertTo-DistinctWordsArray {
+    param($str)
+    # get words array from string
+    $words = $str.Split(' ')
+    # get unique words
+    $words = $words | Get-Unique 
+    return $words
+}
+
+function Edit-AddSearchAttribues
+{
+    param($libraryPaper)
+
+    # add normalized title attribute for title search
+    $normalizedTitle = ConvertTo-NormalizedStr $libraryPaper.OriginalTitle
+    $libraryPaper | Add-Member -MemberType NoteProperty -Name 'Title' -Value  $normalizedTitle
+
+    # add normlized title words for partial title search
+    $titleWords = ConvertTo-DistinctWordsArray $normalizedTitle
+    $libraryPaper | Add-Member -MemberType NoteProperty -Name 'TitleWords' -Value  $titleWords
+}
```

Comment:
Why do you need to do this? You could just request "Ti" and "W" back from MAKES.

The only data you need to persist from the "private" data is just the URL's

Reply:
In case we can't link the paper, we can still support search. Also, this is address your previous feedback:
"It's unclear why you're merging these. The existing presentation title should already be sufficient, plus it is in the correct display form.
It seems like you'd only want to augment fields that aren't present at all"

---
In reply to: [506728642](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/185#discussion_r506728642) [](ancestors = 506728642)